File: Python/Math-Array-String/68.py
# 442. Find All Duplicates in an Array

from typing import List

class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        result = []
        for num in nums:
            index = abs(num) - 1
            if nums[index] < 0:
                result.append(abs(num))
            else:
                nums[index] = -nums[index]

        return result
    

sol = Solution()
output = sol.findDuplicates([1,3,5,4,5])
print(output) # Output: [5]

File: JavaScript/Math-Array-String/68.js
// 442. Find All Duplicates in an Array


/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDuplicates = function(nums) {
    const result = [];
    
    for (let i = 0; i < nums.length; i++) {
        const index = Math.abs(nums[i]) - 1; // Map the value to its corresponding index
        
        if (nums[index] < 0) {
            // If already marked negative, it means this is a duplicate
            result.push(Math.abs(nums[i]));
        } else {
            // Mark the corresponding index as visited by making it negative
            nums[index] = -nums[index];
        }
    }
    
    return result;
};

// Example usage
console.log(findDuplicates([4, 3, 2, 7, 8, 2, 3, 1])); // Output: [2, 3]
console.log(findDuplicates([1, 1, 2]));                // Output: [1]
console.log(findDuplicates([1]));                      // Output: []
